python/Ganga/Utility/execute.py

```diff
@@ -4,36 +4,58 @@
 from Ganga.Utility.logging import getLogger
 logger = getLogger()
 
+#execute_lock = threading.RLock()
+
 def env_update_script(indent=''):
-    fdread, fdwrite = os.pipe()
+    """ This function creates an extension to a python script, or just a python script to be run at the end of the
+    piece of code we're interedted in.
+    This piece of code will dump the environment after the execution has taken place into a temporary file.
+    This returns a tuple of the script it's generated and the NamedTemporaryFile that will be used to store the env
+    Args:
+        indent (str): This is the indent to apply to the script if this script is to be appended to a python file
+    """
+    this_env_file = NamedTemporaryFile(delete=False)
+    with open(this_env_file.name, 'w') as some_file:
+        pass
     this_script = '''
 import os, pickle
-os.close(###FD_READ###)
-with os.fdopen(###FD_WRITE###,'wb') as envpipe:
+with open('###TEMP_WRITE###','w') as envpipe:
     pickle.dump(os.environ, envpipe)
 '''
     from Ganga.GPIDev.Lib.File.FileUtils import indentScript
     script = indentScript(this_script, '###INDENT###')
 
-    script =  script.replace('###INDENT###'  , indent      )\
-                    .replace('###FD_READ###' , str(fdread) )\
-                    .replace('###FD_WRITE###', str(fdwrite))
-    return script, fdread, fdwrite
+    script =  script.replace('###INDENT###'    , indent            )\
+                    .replace('###TEMP_WRITE###', this_env_file.name)
+
+    return script, this_env_file
 
 # /\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
 
 
 def python_wrapper(command, python_setup='', update_env=False, indent=''):
-    fdread, fdwrite = os.pipe()
+    """ This section of code wraps the given python command inside a small wrapper class to allow us to control the output.
+    Optionally we can also append to the end of this file a script to allow us to extract the environment after we've
+    finished executing our command.
+    Args:
+        command (str): This is the pthon code to be executed (can be multi-line)
+        python_setup (str): This is some python code to be executed before the python code in question (aka a script header.
+        update_env (bool): Contol whether we want to capture the env after running
+        indent (str): This allows for an indent to be applied to the script so it can be placed inside other python scripts
+    This returns the NamedTemporaryFile objects for the env_update_script, the python wrapper itself and the script which has been generated to be run
+    """
+    this_pkl_file = NamedTemporaryFile(delete=False)
+    with open(this_pkl_file.name, 'w') as some_file:
+        pass
     this_script = '''
 from __future__ import print_function
 import os, sys, pickle, traceback
-os.close(###PKL_FDREAD###)
-with os.fdopen(###PKL_FDWRITE###, 'wb') as PICKLE_STREAM:
+with open('###TEMP_WRITE###', 'w') as PICKLE_STREAM:
```

Comment:
I'm not going to push the change as I too found it more difficult to debug using pipes. I'm just very wary about creating too many temporary files. Especially as one will be made for every `execute(shell=False)` call, just to get the output. This will be done for every single DIRAC call I believe.